In [1]:
import pandas as pd

In [2]:
df=pd.read_csv("data.csv")
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St    2.0    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St    2.0    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St    3.0    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St    3.0    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

1--no. of rows and no. of columns in the dataset.

In [3]:
print("No of row and column :", df.shape)

No of row and column : (23678, 21)


2--the datatypes of each column

In [4]:
df.dtypes

Suburb            object
Address           object
Rooms            float64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

3--the row indexes of rows having all null values

In [5]:
data=df.isnull().sum(axis=1)[df.isnull().sum(axis=1)==21]
data

31       21
1145     21
1348     21
2119     21
3277     21
         ..
22288    21
22704    21
22918    21
23153    21
23341    21
Length: 131, dtype: int64

In [6]:
df.drop((data),inplace=True)

In [7]:
df.shape

(23677, 21)

5--column names having missing values

In [8]:
df.isnull().sum()

Suburb             131
Address            131
Rooms              131
Type               131
Price             5282
Method             131
SellerG            131
Date               131
Distance           132
Postcode           132
Bedroom2          4611
Bathroom          4614
Car               4756
Landsize          6267
BuildingArea     13659
YearBuilt        12137
CouncilArea       8021
Lattitude         4434
Longtitude        4434
Regionname         132
Propertycount      132
dtype: int64

6--missing values and their percentage in each column.

In [9]:
df.isnull().mean().round(4) * 100


Suburb            0.55
Address           0.55
Rooms             0.55
Type              0.55
Price            22.31
Method            0.55
SellerG           0.55
Date              0.55
Distance          0.56
Postcode          0.56
Bedroom2         19.47
Bathroom         19.49
Car              20.09
Landsize         26.47
BuildingArea     57.69
YearBuilt        51.26
CouncilArea      33.88
Lattitude        18.73
Longtitude       18.73
Regionname        0.56
Propertycount     0.56
dtype: float64

7--the row indexes having missing values

In [10]:
df.isnull().sum(axis=1)

0        3
1        2
2        0
3        3
4        0
        ..
23673    2
23674    8
23675    4
23676    1
23677    2
Length: 23677, dtype: int64

8--the missing values and their percentage in each row

In [11]:
percent_missing = df.isna().sum(axis=1) * 100 / len(df)

In [12]:
percent_missing

0        0.012671
1        0.008447
2        0.000000
3        0.012671
4        0.000000
           ...   
23673    0.008447
23674    0.033788
23675    0.016894
23676    0.004224
23677    0.008447
Length: 23677, dtype: float64

9--the column names having more than 30% missing values

In [13]:
df.isnull().sum()>0.30*len(df)

Suburb           False
Address          False
Rooms            False
Type             False
Price            False
Method           False
SellerG          False
Date             False
Distance         False
Postcode         False
Bedroom2         False
Bathroom         False
Car              False
Landsize         False
BuildingArea      True
YearBuilt         True
CouncilArea       True
Lattitude        False
Longtitude       False
Regionname       False
Propertycount    False
dtype: bool

In [14]:
df.drop(["BuildingArea","YearBuilt","CouncilArea"],axis=1,inplace=True)

10--the number and percentage of rows having more than 5 missing values. If the percentage of such rows is less than 20% then remove all these rows.

In [15]:
df.isnull().sum()

Suburb            131
Address           131
Rooms             131
Type              131
Price            5282
Method            131
SellerG           131
Date              131
Distance          132
Postcode          132
Bedroom2         4611
Bathroom         4614
Car              4756
Landsize         6267
Lattitude        4434
Longtitude       4434
Regionname        132
Propertycount     132
dtype: int64

In [16]:
for x in df.columns:
    if df[x].isnull().sum()<0.2*len(df):
        df.drop([x],axis=1,inplace=True)

In [17]:
df.head()

Price  Car  Landsize
0        NaN  1.0     126.0
1  1480000.0  1.0     202.0
2  1035000.0  0.0     156.0
3        NaN  1.0       0.0
4  1465000.0  0.0     134.0

11--the description of data and percentage of missing values still left column wise.

In [18]:
# description
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23677 entries, 0 to 23677
Data columns (total 3 columns):
Price       18395 non-null float64
Car         18921 non-null float64
Landsize    17410 non-null float64
dtypes: float64(3)
memory usage: 739.9 KB


In [19]:
# percentage of missing value
(df.isnull().sum()/len(df))*100

Price       22.308569
Car         20.087004
Landsize    26.468725
dtype: float64

12---explanation and treat the missing values in 'Price' column.

Following are the step:
    1-Importing imputer from sklearn.preprocessing 
    2-filling nan value with mean value

In [20]:
from sklearn.preprocessing import Imputer

In [21]:
y=Imputer(missing_values="NaN",strategy="mean")

C:\Users\Admin\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:66: DeprecationWarning: Class Imputer is deprecated; Imputer was deprecated in version 0.20 and will be removed in 0.22. Import impute.SimpleImputer from sklearn instead.
  warnings.warn(msg, category=DeprecationWarning)


In [22]:
df["Price"]=y.fit_transform(df["Price"].values.reshape(-1,1))

In [23]:
df.isnull().sum()

Price          0
Car         4756
Landsize    6267
dtype: int64

13--the percentage of missing values in each column

In [24]:
(df.isnull().sum()/len(df))*100

Price        0.000000
Car         20.087004
Landsize    26.468725
dtype: float64

In [25]:
df["Car"]=df["Car"].fillna(df["Car"].mean())
df["Landsize"]=df["Landsize"].fillna(df["Landsize"].mean())

In [26]:
df.isnull().sum()

Price       0
Car         0
Landsize    0
dtype: int64

14--the number and percentage of columns removed

In [27]:
df.shape

(23677, 3)

In [28]:
# we can see now there is only 3 column

print("total no of column removed :",(21-3))
print("total percentage of removed colunms : ",(18/21)*100)

total no of column removed : 18
total percentage of removed colunms :  85.71428571428571


15--The amount of data lost is more than 85%.We cannot predict price of house because there is very less information left.Although if we try we will not get very accurate result.There are many important column such as (rooms,type,distance,bathroom,year built) are drop in cleaning process.